doc/mojitos.1
.Dd January 9, 2023
.Dt MOJITOS 1
.Os
.Sh NAME
.Nm mojitos
.Nd An open source system, energy and network monitoring tool.
.Sh SYNOPSIS
.Nm mojitos
.Op Fl rsu
.Op Fl t Ar time
.Op Fl f Ar freq
.Op Fl p Ar perf_list
.Op Fl d Ar net_device
.Op Fl o Ar logfile
.Op Fl e Ar cmd ...
.Nm mojitos
.Op Fl l
.Sh DESCRIPTION
.Nm
enables monitoring the system, its energy comsumption and the network activity, at the OS level.
It runs on GNU/Linux.
.Pp
.Nm
supports the following options:
.Bl -tag -width Ds
.It Fl s
Enable overhead statistics (in nanoseconds).
.It Fl u
Enable system-level load monitoring.
.It Fl r
Enable RAPL.
.It Fl p Ar perf_list
Enable performance counters.
The argument is a coma separated list of performance counters.
.It Fl d Ar net_device
Enable network monitoring.
.It Fl l
List the available performance counters and quit.
.It Fl t Ar time
Set duration value (in seconds). If 0, then loops indefinitely.
.It Fl f Ar freq
Set amount of measurements per second.
.It Fl e Ar cmd ...
Execute a command with optional arguments.
If this option is used, any usage of -t or -f is ignored.
.El
.Sh EXIT STATUS
.Ex
.Sh EXAMPLES
RAPL values during 2 seconds with a frequency of 2Hz:
.Bd -literal -offset indent
$ mojitos -t 2 -f 2 -r
#timestamp package-00 core0 dram0
1036389.135659868 10986 2869 1526
1036389.500183551 1291440 255736 515562
1036390.000754048 1333553 228393 689513
1036390.500113978 1581967 267944 701536
.Ed
.Pp
Performance counters (cpu_cycle, cache_ll_r_a and page_faults) during 4 seconds
with a frequency of 1Hz.
In the context of cache performance counters (as with "cache_ll_r_a"),
 _r and _w suffixes mean respectively read and write,
while the _a, _m and _p ones mean respectively access, miss and pending.
.Pp
.Bd -literal -offset indent
$ ./mojitos -t 4 -f 1 -p cpu_cycles,cache_ll_r_a,page_faults
#timestamp cpu_cycles cache_ll page_faults
1036846.351749455 571199 1232 0
1036847.001098880 348173344 2451387 872
1036848.000166158 388112961 2509305 791
1036849.000191883 402255979 2625283 799
.Ed
.Pp
Network values with no time limit with a frequency of 1Hz.
rxp and txp are the number of received and sent packets, while rxb and txp are
the number of received and sent bytes.
.Pp
.Bd -literal -offset indent
$ ./mojitos -t 0 -f 1 -d enp0s25
#timestamp rxp rxb txp txb
1036559.277376027 0 0 0 0
1036560.000161101 4 581 2 179
1036561.000083968 178 268675 55 4954
1036562.000076162 11 1010 5 510
1036563.000069724 17 1643 12 3602
1036564.000113394 990 1493008 369 27299
.Ed
.Pp
overhead of the monitoring when using RAPL and the cpu_cycle performance counter:
.Pp
.Bd -literal -offset indent
$ ./mojitos -t 5 -f 1 -p cpu_cycles -r -s
#timestamp cpu_cycles package-00 core0 dram0 overhead
1036988.197227391 162214 19898 4944 1586 149612
1036989.000151326 332613664 2513116 379577 1115171 739573
1036990.000116433 482150700 3321341 587218 1380673 315719
1036991.000182835 525984292 3592582 691221 1385982 272182
1036992.000165117 397678789 2770561 444030 1375729 510379
.Ed
.Sh SEE ALSO
.Xr perf 1 ,
.Xr lscpu 1 ,
.Xr sysfs 5 ,
.Xr proc 5
.Sh LICENSE
MojitO/S is published under the GPL3 license and is part of the
.Lk https://www.irit.fr/energumen/ [Energumen Project]
.Sh BUGS
Yes.


src/mojitos.c
/*******************************************************

 Copyright (C) 2018-2019 Georges Da Costa <georges.da-costa@irit.fr>

    This file is part of Mojitos.

    Mojitos is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    Mojitos is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with MojitO/S.  If not, see <https://www.gnu.org/licenses/>.

 *******************************************************/

#include <inttypes.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <unistd.h>

#define NB_MAX_OPTS 10
#define NB_MAX_CAPTORS 20

#define OPTPARSE_IMPLEMENTATION
#define OPTPARSE_API static
#include "optparse.h"

#define UNUSED(expr) do { (void)(expr); } while (0)
#define PANIC(code, fmt, ...)  				 \
	do {									 \
		fprintf(stderr, "Exit on error: ");  \
		fprintf(stderr, fmt, ##__VA_ARGS__); \
		fprintf(stderr, "\n"); 			     \
		exit(code); 						 \
	} while (0)

typedef unsigned int (*initializer_t)(char *, void **);
typedef void (*labeler_t)(char **, void *);
typedef unsigned int (*getter_t)(uint64_t *, void *);
typedef void (*cleaner_t)(void *);

struct captor {
    char *usage_arg;
    char *usage_msg;
    initializer_t init;
    getter_t get;
    cleaner_t clean;
    labeler_t label;
};

struct captor captors[NB_MAX_CAPTORS];

struct optparse_long longopts[NB_MAX_OPTS + NB_MAX_CAPTORS] = {
    {"overhead-stats", 's', OPTPARSE_NONE},
    {"list", 'l', OPTPARSE_NONE},
    {"freq", 'f', OPTPARSE_REQUIRED},
    {"time", 't', OPTPARSE_REQUIRED},
    {"exec", 'e', OPTPARSE_REQUIRED},
    {"logfile", 'o', OPTPARSE_REQUIRED},
};

int nb_defined_captors = 0;

#include "captors.h"

void usage(char **argv)
{
    printf("Usage : %s [OPTIONS] [CAPTOR ...] [-o logfile] [-e cmd ...]\n"
           "\nOPTIONS:\n"
           "-t <time>\n"
           "-f <freq>\n"
           "-l\t\tlist the possible performance counters and quit\n"
           "-s\t\tenable overhead statistics in nanoseconds\n"
           "if time==0 then loops infinitively\n"
           "if -e is present, time and freq are not used\n"
           , argv[0]);

    if (nb_defined_captors == 0) {
        // no captor to show
        exit(EXIT_FAILURE);
    }

    printf("\nCAPTORS:\n");

    for (int i = 0; i < nb_defined_captors; i++) {
        printf("-%c", longopts[NB_MAX_OPTS + i].shortname);
        if (captors[i].usage_arg != NULL) {
            printf(" %s", captors[i].usage_arg);
        }
        printf("\n\t%s\n", captors[i].usage_msg);
    }

    exit(EXIT_FAILURE);
}

void sighandler(int none)
{
    UNUSED(none);
}

void flush(int none)
{
    UNUSED(none);
    exit(0);
}

FILE *output;
void flushexit()
{
    if (output != NULL) {
        fflush(output);
        fclose(output);
    }
}

unsigned int nb_sources = 0;
void **states = NULL;
getter_t *getter = NULL;
cleaner_t *cleaner = NULL;

unsigned int nb_sensors = 0;
char **labels = NULL;
uint64_t *values = NULL;

void add_source(struct captor *cpt, char *arg)
{
    nb_sources++;
    initializer_t init = cpt->init;
    labeler_t labeler = cpt->label;
    getter_t get = cpt->get;
    cleaner_t clean = cpt->clean;

    states = realloc(states, nb_sources * sizeof(void *));
    int nb = init(arg, &states[nb_sources - 1]);

    if (nb == 0) {
        nb_sources--;
        states = realloc(states, nb_sources * sizeof(void *));
        return;
    }

    getter = realloc(getter, nb_sources * sizeof(void *));
    getter[nb_sources - 1] = get;
    cleaner = realloc(cleaner, nb_sources * sizeof(void *));
    cleaner[nb_sources - 1] = clean;

    labels = realloc(labels, (nb_sensors + nb) * sizeof(char *));
    labeler(labels + nb_sensors, states[nb_sources - 1]);

    values = realloc(values, (nb_sensors + nb) * sizeof(uint64_t));
    nb_sensors += nb;
}

int main(int argc, char **argv)
{
    int total_time = 1;
    int delta = 0;
    int frequency = 1;
    char **application = NULL;
    int stat_mode = -1;

    init_captors();

    if (argc == 1) {
        usage(argv);
    }

    output = stdout;

    atexit(flushexit);
    signal(15, flush);

    int opt;
    struct optparse options;
    options.permute = 0;

    optparse_init(&options, argv);
    while ((opt = optparse_long(&options, longopts, NULL)) != -1 && application == NULL) {
        switch (opt) {
        case 'f':
            frequency = atoi(options.optarg);
            break;
        case 't':
            total_time = atoi(options.optarg);
            delta = 1;
            if (total_time == 0) {
                total_time = 1;
                delta = 0;
            }
            break;
        case 's':
            stat_mode = 0;
            break;
        case 'l':
            show_all_counters();
            exit(EXIT_SUCCESS);
        case 'o':
            if ((output = fopen(options.optarg, "wb")) == NULL) {
                perror("fopen");
                PANIC(1, "-o %s", options.optarg);
            }
            break;
        case 'e':
            application = options.argv;
            signal(17, sighandler);
            break;
        default: {
            int ismatch = 0;
            for (int i = 0; i < nb_defined_captors && !ismatch; i++) {
                if (opt == longopts[NB_MAX_OPTS + i].shortname) {
                    ismatch = 1;
                    add_source(&captors[i], options.optarg);
                }
            }
            if (!ismatch) {
                fprintf(stderr, "%s: %s\n", argv[0], options.errmsg);
                usage(argv);
            }
        }
        }
    }

    setvbuf(output, NULL, _IONBF, BUFSIZ);
    struct timespec ts;
    struct timespec ts_ref;

    fprintf(output, "#timestamp ");

    for (unsigned int i = 0; i < nb_sensors; i++) {
        fprintf(output, "%s ", labels[i]);
    }

    if (stat_mode == 0) {
        fprintf(output, "overhead ");
    }

    fprintf(output, "\n");

    unsigned long int stat_data = 0;

    for (int temps = 0; temps < total_time * frequency; temps += delta) {
        clock_gettime(CLOCK_MONOTONIC, &ts_ref);

        // Get Data
        unsigned int current = 0;

        for (unsigned int i = 0; i < nb_sources; i++) {
            current += getter[i](&values[current], states[i]);
        }

        if (application != NULL) {

            if (fork() == 0) {
                execvp(application[0], application);
                exit(0);
            }

            pause();
            clock_gettime(CLOCK_MONOTONIC, &ts);

            if (ts.tv_nsec >= ts_ref.tv_nsec) {
                fprintf(output, "%ld.%09ld ", (ts.tv_sec - ts_ref.tv_sec), ts.tv_nsec - ts_ref.tv_nsec);
            } else {
                fprintf(output, "%ld.%09ld ", (ts.tv_sec - ts_ref.tv_sec) - 1, 1000 * 1000 * 1000 + ts.tv_nsec - ts_ref.tv_nsec);
            }
        } else {
#ifdef DEBUG
            clock_gettime(CLOCK_MONOTONIC, &ts);
            fprintf(stderr, "%ld\n", (ts.tv_nsec - ts_ref.tv_nsec) / 1000);
            //Indiv: mean: 148 std: 31 % med: 141 std: 28 %
            //Group: mean: 309 std: 41 % med: 297 std: 39 %
#endif

            if (stat_mode == 0) {
                clock_gettime(CLOCK_MONOTONIC, &ts);

                if (ts.tv_nsec >= ts_ref.tv_nsec) {
                    stat_data = ts.tv_nsec - ts_ref.tv_nsec;
                } else {
                    stat_data = 1000 * 1000 * 1000 + ts.tv_nsec - ts_ref.tv_nsec;
                }
            }

            // Treat Data
            fprintf(output, "%ld.%09ld ", ts_ref.tv_sec, ts_ref.tv_nsec);
        }

        for (unsigned int i = 0; i < nb_sensors; i++) {
            /* "PRIu64" is a format specifier to print uint64_t values */
            fprintf(output, "%" PRIu64 " ", values[i]);
        }

        if (stat_mode == 0) {
            fprintf(output, "%ld ", stat_data);
        }

        fprintf(output, "\n");

        if (application != NULL) {
            break;
        }

        clock_gettime(CLOCK_MONOTONIC, &ts);
        usleep(1000 * 1000 / frequency - (ts.tv_nsec / 1000) % (1000 * 1000 / frequency));
    }

    for (unsigned int i = 0; i < nb_sources; i++) {
        cleaner[i](states[i]);
    }

    if (nb_sources > 0) {
        free(getter);
        free(cleaner);
        free(labels);
        free(values);
        free(states);
    }
}

